### PR TITLE
Migrating types command to the newshell

### PR DESCRIPTION
**Pull request by XVilka**
# **DO NOT SQUASH**

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

- Migrating `t` commands to the newshell and fixes along the way.
- Removed `touch` command
- Added `tek` command to print enum contents in the SDB format
- Added `tef` command to search for the matching enums by the value (necessary for https://github.com/rizinorg/rizin/issues/697)
- Removed `tfc` command and now `tf` behaves as one - it prints the function types in C format
- Moved old `tf` command to `tfk` to print SDB format

For example, lets define two enums:
```
td "enum BLA { A = 1, B = 0x80, C = 2 };"
td "enum FOO { G = 2, B = 0x90, P = 3254543};"
```
If you run `tef 2` it will show:
```
BLA.C
FOO.G
```
If you run `tef 1` it will show:
```
BLA.A
```

**Test plan**

All green

### CODE REVIEW

File: librz/core/cmd_type.c [hunk 1]
```diff
@@ -254,6 +253,163 @@ static void types_cc_print(RzCore *core, const char *cc, RzOutputMode mode) {
...
+			rz_list_foreach (list, iter, member) {
+				pj_kn(pj, member->name, rz_num_math(NULL, member->val));
+			}
+			pj_end(pj);
+		}
+		pj_end(pj);
+		rz_cons_printf("%s\n", pj_string(pj));
+		pj_free(pj);
+		rz_list_free(list);
+	} break;
```
Comment (ret2libc) [inline librz/core/cmd_type.c]: put the break inside the  `{ .. }`. At least it will be with the right indentation.

File: test/db/cmd/types [hunk 2]
```diff
@@ -278,18 +278,108 @@ RUN
...
+RUN
+
+NAME=tej
+FILE=-
+CMDS=<<EOF
+td "enum FOO { A=1, B=0x100, C=0xFF, D=4, E=5 };"
+td "enum BAR { G=0x800, Z=0xEE, P=4, E=5 };"
+tej FOO
+tej BAR
+tej
```
Comment (ret2libc) [inline test/db/cmd/types:316]: this should output an array, not just 2 objects, I think.

File: librz/core/cmd_type.c [hunk 3]
```diff
@@ -1115,7 +1209,7 @@ static void type_list_c_all(RzCore *core) {
 	// TODO: Change the logic maybe?
 	//eprintf("Specify the type");
 	//return RZ_CMD_STATUS_ERROR;
-	rz_core_cmd0(core, "tfc");
+	rz_core_cmd0(core, "tfc"); // seems return nothing
```
Comment (ret2libc) [inline librz/core/cmd_type.c]: use `types_function_print`?

File: librz/core/cmd_type.c [hunk 4]
```diff
@@ -1883,6 +1919,84 @@ RZ_IPI RzCmdStatus rz_type_define_handler(RzCore *core, int argc, const char **a
...
+	// TODO: Reconsider the `te <enum_name> <member_value>` syntax change
+	const char *member_value = argc > 2 ? argv[2] : NULL;
+	if (enum_name) {
+		if (member_value) {
+			return types_enum_member_find(core, enum_name, member_value);
+		} else {
+			types_enum_print(core, enum_name, mode);
+		}
+	} else {
+		// A special case, since by default `te` returns only the list of all enums
```
Comment (ret2libc) [inline librz/core/cmd_type.c:1976]: can't we just add support for `te` and `teq`?

File: librz/core/cmd_type.c [hunk 5]
```diff
@@ -1883,6 +1919,84 @@ RZ_IPI RzCmdStatus rz_type_define_handler(RzCore *core, int argc, const char **a
 	return RZ_CMD_STATUS_OK;
 }
 
+RZ_IPI RzCmdStatus rz_type_list_enum_handler(RzCore *core, int argc, const char **argv, RzOutputMode mode) {
+	const char *enum_name = argc > 1 ? argv[1] : NULL;
+	// TODO: Reconsider the `te <enum_name> <member_value>` syntax change
```
Comment (ret2libc) [inline librz/core/cmd_type.c:1962]: This is something that always bugs me... Having the same handler that does both listing and addition. I'm still not sure if it's a real problem or not and, if it is, how to improve the situation... Adding yet another command is a bit meh.

File: librz/core/cmd_type.c [hunk 6]
```diff
@@ -1883,6 +1919,84 @@ RZ_IPI RzCmdStatus rz_type_define_handler(RzCore *core, int argc, const char **a
...
+RZ_IPI RzCmdStatus rz_type_enum_find_handler(RzCore *core, int argc, const char **argv) {
+	const char *enum_value = argc > 1 ? argv[1] : NULL;
+	return types_enum_member_find_all(core, enum_value);
+}
+
+RZ_IPI RzCmdStatus rz_type_list_function_handler(RzCore *core, int argc, const char **argv, RzOutputMode mode) {
+	const char *function = argc > 1 ? argv[1] : NULL;
+	Sdb *TDB = core->analysis->sdb_types;
+	if (function) {
+		// A special case, since by default `tf name` returns the output in SDB format
```
Comment (ret2libc) [inline librz/core/cmd_type.c]: Same here. Can't we just support both STANDARD and SDB modes? For consistency, sdb output should be printed with `tfk`

Comment (ret2libc) [reply librz/core/cmd_type.c]: I'm trying the commands, and I feel like `tfc` should really be just `tf` and `tf` should become `tfk`. THis would be consistent with other commands.

File: librz/core/cmd_type.c [hunk 7]
```diff
@@ -1883,6 +1919,84 @@ RZ_IPI RzCmdStatus rz_type_define_handler(RzCore *core, int argc, const char **a
...
+		if (mode == RZ_OUTPUT_MODE_JSON) {
+			print_keys(TDB, core, stdiffunc, printfunc_json_cb, false);
+		} else {
+			print_keys(TDB, core, stdiffunc, printkey_cb, false);
+		}
+	}
+	return RZ_CMD_STATUS_OK;
+}
+
+RZ_IPI RzCmdStatus rz_type_function_c_handler(RzCore *core, int argc, const char **argv, RzOutputMode mode) {
```
Comment (ret2libc) [inline librz/core/cmd_type.c]: in oldshell i'm not able to make `tfcj` work. Also, what would be the difference between `tfcj` and `tcj` in this new implementation? I think they are the same thing. Honestly, I can't think about a "C JSON" reprensetation. IMHO this should be merged with `tf`, as said above.

File: librz/core/cmd_descs/cmd_type.yaml [hunk 8]
```diff
@@ -81,6 +150,28 @@ commands:
         cname: type_noreturn_del_all
         summary: Remove all noreturn references
         args: []
+  - name: to
```
Comment (ret2libc) [inline librz/core/cmd_descs/cmd_type.yaml:143]: Instead of using `load` as cname and `Load` as summary for most of these, i would rephrase a bit and use the word `open`/`Open` because i guess `to` stands for `type open`, meaning that you open a file to work on types. `to` opens a file and loads types from it, `toe` open the editor, `tos` open a sdb file and load types from it.

File: librz/core/cmd_type.c [hunk 9]
```diff
@@ -1933,6 +2047,24 @@ RZ_IPI RzCmdStatus rz_type_noreturn_del_all_handler(RzCore *core, int argc, cons
...
+
+RZ_IPI RzCmdStatus rz_type_load_editor_handler(RzCore *core, int argc, const char **argv) {
+	const char *typename = argc > 1 ? argv[1] : NULL;
+	types_load_editor(core, typename);
+	return RZ_CMD_STATUS_OK;
+}
+
+RZ_IPI RzCmdStatus rz_type_load_sdb_handler(RzCore *core, int argc, const char **argv) {
+	const char *path = argc > 1 ? argv[1] : NULL;
+	types_load_sdb(core, path);
```
Comment (ret2libc) [inline librz/core/cmd_type.c]: just do `types_load_sdb(core, argv[1])` and remove the previous line. This argument is anyway required.

File: librz/core/cmd_descs/cmd_type.yaml [hunk 10]
```diff
@@ -81,6 +150,28 @@ commands:
...
+    summary: Load types from C header file
+    subcommands:
+      - name: to
+        cname: type_load_file
+        summary: Load types from C header file
+        args:
+          - name: file
+            type: RZ_CMD_ARG_TYPE_STRING
+      - name: toe
```
Comment (ret2libc) [inline librz/core/cmd_descs/cmd_type.yaml:152]: Other places use `!` as suffix for the editor, however i think there's no real consistent way at the moment. I like `e`, but it could be some commands already have it occupied... Maybe `E`? Not sure...

File: librz/core/cmd_type.c [hunk 11]
```diff
@@ -1153,6 +1247,78 @@ static void type_define(RzCore *core, const char *type) {
...
+			fprintf(stderr, "%s", error_msg);
+			free(error_msg);
+		}
+	}
+	free(homefile);
+}
+
+static void types_load_editor(RzCore *core, const char *typename) {
+	Sdb *TDB = core->analysis->sdb_types;
+	char *str = rz_core_cmd_strf(core, "tc %s", typename);
```
Comment (ret2libc) [inline librz/core/cmd_type.c]: `typename` might be null when called from the newshell handler, because the argument is optional. I think this is not alright.

File: librz/core/cmd_type.c [hunk 12]
```diff
@@ -1933,6 +2047,24 @@ RZ_IPI RzCmdStatus rz_type_noreturn_del_all_handler(RzCore *core, int argc, cons
 	return RZ_CMD_STATUS_OK;
 }
 
+RZ_IPI RzCmdStatus rz_type_load_file_handler(RzCore *core, int argc, const char **argv) {
+	const char *path = argc > 1 ? argv[1] : NULL;
+	types_load_file(core, path);
```
Comment (ret2libc) [inline librz/core/cmd_type.c]: just do `types_load_file(core, argv[1]);`

File: librz/core/cmd_type.c (hunk 6 above)
Comment (XVilka) [reply librz/core/cmd_type.c]: Agree on this.

File: librz/core/cmd_descs/cmd_type.yaml (hunk 10 above)
Comment (XVilka) [reply librz/core/cmd_descs/cmd_type.yaml:152]: I can note it in my TODO but this is out of the scope of this PR, imho.

File: librz/core/cmd_type.c (hunk 3 above)
Comment (XVilka) [reply librz/core/cmd_type.c]: Done.

File: librz/core/cmd_type.c (hunk 6 above)
Comment (XVilka) [reply librz/core/cmd_type.c]: Fixed in https://github.com/rizinorg/rizin/pull/728/commits/636b315bebfa6e1e461c0c9910d8ea1b9ddc77a5

File: librz/core/cmd_type.c (hunk 7 above)
Comment (XVilka) [reply librz/core/cmd_type.c]: Removed it.

File: librz/core/cmd_type.c (hunk 4 above)
Comment (XVilka) [reply librz/core/cmd_type.c:1976]: I am not sure about this to be honest. It makes sense to list only enum names by default, since there can be many.

File: librz/core/cmd_type.c (hunk 5 above)
Comment (XVilka) [reply librz/core/cmd_type.c:1962]: It doesn't add the value, it can only be done through `td` or manually in SDB. But it gives fine grained access for the enum members.

File: librz/core/cmd_descs/cmd_type.yaml (hunk 10 above)
Comment (ret2libc) [reply librz/core/cmd_descs/cmd_type.yaml:152]: It was just for discussion, to check if you have any idea about it.

File: test/db/cmd/types (hunk 2 above)
Comment (ret2libc) [reply test/db/cmd/types:316]: @XVilka seen this?

File: librz/core/cmd_descs/cmd_type.yaml [hunk 13]
```diff
@@ -50,6 +50,65 @@ commands:
...
+          - name: field
+            type: RZ_CMD_ARG_TYPE_STRING
+      - name: tec
+        cname: type_enum_c
+        summary: Show enum in the C output format
+        args:
+          - name: enum
+            type: RZ_CMD_ARG_TYPE_STRING
+            optional: true
+      - name: ted
```
Comment (ret2libc) [inline librz/core/cmd_descs/cmd_type.yaml:85]: this should probably be a subcommand of `tec` (e.g. `tecn`, because also `?e` has `?en` to avoid newlines).

File: librz/core/cmd_type.c (hunk 4 above)
Comment (ret2libc) [reply librz/core/cmd_type.c:1976]: Right, but maybe we can consider STANDARD just the enum names and LONG the C version. This way you'd have just one command to list enums in different formats:
`te` -> just enum names (STANDARD)
`tej` -> enum in json (JSON)
`tek` -> enum in sdb (SDB)
`tel` -> enum in C (LONG)

File: librz/core/cmd_descs/cmd_type.yaml (hunk 13 above)
Comment (XVilka) [reply librz/core/cmd_descs/cmd_type.yaml:85]: We have the same pattern for `tsd`, `tud`, etc. Do you suggest to convert all of them?

File: librz/core/cmd_type.c (hunk 4 above)
Comment (XVilka) [reply librz/core/cmd_type.c:1976]: But then `te <enum name>` command wouldn't have any meaning like right now. Other types have output with `pf` string when you call e.g. `tu <union name>`.

File: test/db/cmd/types (hunk 2 above)
Comment (XVilka) [reply test/db/cmd/types:316]: It will require passing `PJ` pointer around :(

File: librz/core/cmd_descs/cmd_type.yaml [hunk 14]
```diff
@@ -81,6 +140,28 @@ commands:
...
         args: []
+  - name: to
+    summary: Open C header file and load types from it
+    subcommands:
+      - name: to
+        cname: type_open_file
+        summary: Open C header file and load types from it
+        args:
+          - name: file
+            type: RZ_CMD_ARG_TYPE_STRING
```
Comment (ret2libc) [inline librz/core/cmd_descs/cmd_type.yaml]: TYPE_FILE

File: librz/core/cmd_descs/cmd_type.yaml [hunk 15]
```diff
@@ -81,6 +140,28 @@ commands:
...
+        args:
+          - name: type
+            type: RZ_CMD_ARG_TYPE_STRING
+            optional: true
+      - name: tos
+        cname: type_open_sdb
+        summary: Open SDB file and load types from it
+        args:
+          - name: file
+            type: RZ_CMD_ARG_TYPE_STRING
```
Comment (ret2libc) [inline librz/core/cmd_descs/cmd_type.yaml]: type file

File: test/db/cmd/types (hunk 2 above)
Comment (ret2libc) [reply test/db/cmd/types:316]: it is necessary, I think. If you have a json command that returns more than one object, they should be wrapped in `[...]`.

File: librz/core/cmd_type.c (hunk 4 above)
Comment (ret2libc) [reply librz/core/cmd_type.c:1976]: So many inconsistencies :( I even wonder why one would want to print "pf format" instead of something easier... I like more and more the idea of having `:<mode>` as a way to specify the output mode you want. So commands could just have `<cmd>:pf` or `<cmd>:sdb` `<cmd>:json` instead of having to find a letter for each case, ending up with many inconsistencies :(

Comment (XVilka) [reply librz/core/cmd_type.c:1976]: Yes, the idea of having the ":<mode>" is the good one.

File: librz/core/cmd_type.c [hunk 16]
```diff
@@ -321,14 +319,23 @@ static void types_enum_print_all(RzCore *core, RzOutputMode mode) {
 	Sdb *TDB = core->analysis->sdb_types;
 	SdbKv *kv;
 	SdbListIter *iter;
+	PJ *pj = (mode == RZ_OUTPUT_MODE_JSON) ? pj_new() : NULL;
```
Comment (ret2libc) [inline librz/core/cmd_type.c]: rz_core_pj_new @XVilka

File: test/db/cmd/types (hunk 2 above)
Comment (XVilka) [reply test/db/cmd/types:316]: Done, albeit in an ugly way.

**Review by ret2libc — APPROVED** [review]
I'm ok with this, however please consider the various suggestions once you refactor further the whole types thing.
